app/src/semaphore.h

```diff
@@ -172,7 +172,27 @@ class Semaphore {
     return WaitForSingleObject(semaphore_, milliseconds) == 0;
 #else  // not windows and not mac - should be Linux.
     timespec t = internal::MsToAbsoluteTimespec(milliseconds);
-    return sem_timedwait(semaphore_, &t) == 0;
+    while (true) {
+      int result = sem_timedwait(semaphore_, &t);
+      if (result == 0) {
+        // Return success, since we successfully locked the semaphore.
+        return true;
+      }
+      switch (errno) {
+        case EINTR:
+          // Restart the wait because we were woken up spuriously.
+          continue;
+        case ETIMEDOUT:
+          // Return failure, since the timeout expired.
+          return false;
+        case EINVAL:
+          assert("sem_timedwait() failed with EINVAL" == 0);
```

Comment:
This seems to be a new style of asserting. This assert message doesn't get logged, correct?

Reply:
I saw this assert style somewhere else (I forget where exactly). This assert will fail unconditionally, since the string argument is _never_ equal to `0`. When this style of `assert` fails, sometimes the _only_ trace of it that you see in the log is the text between the assert's `(` and `)`. On several occasions I've had to track down which assertion failed by searching through the code base for the exact text specified to `assert`. So with this style it is easy to track it down and know exactly what the problematic value was that was returned from `sem_timedwait`.